V2

judofyr opened the pull request:
V2 is a rewrite of the parser algorithm based on the ideas presented in https://www.sanity.io/blog/why-we-wrote-yet-another-parser-compiler.

Milestones:

- [x] Implement P1: A recognizer (i.e. it only returns invalid-parse/valid-parse and not a full parse tree) which uses the new formalism
- [ ] Implement P2: A parser which uses the new formalism
- [ ] Implement a P2-to-JS generator

judofyr committed:
Start work on new implementation:

- Introduce P1, a recognizer-only implementation
- Move grammar DSL into Glush::DSL
- Rename "Pattern" to "Expr"
- Remove a lot of existing code which doesn't work yet

judofyr committed:
Improve P1:

- Introduce a new, nested, model for marks
- Add more failing test
- Restructure the algorithm to fix failing tests

judofyr committed:
Use curve_fit for fitting curve

judofyr committed:
Implement new approach for P1

judofyr committed:
P1: Simplify

elimisteve commented:
> For now it’s recommended to wait until the new implementation stabilizes and as such there is limited documentation and examples.

> [I'm going to focus more on it in the following weeks/months, but I can't really give any concrete ETA at the moment.](https://github.com/judofyr/glush/issues/1#issuecomment-615761914)

Do these still apply? Glush looks awesome and I'd love to use it for an ambitious language-to-language translation project, thanks!

judofyr commented:
Sorry for the late response.

I've been looking into Glush lately, but it's been a bit tricky to find an approach which works well. I have some some progress here and there, but nothing significant recently.

judofyr committed:
Lots of changes

- Add back Mark expressions
- Implement an P2 parser
- Add dot-output for the P2 states

judofyr committed:
Add support for next_token

judofyr committed:
Fix inv of equal

judofyr committed:
P2: Small fixes

judofyr committed:
P2: Slight performance improvements

judofyr committed:
P2: Pass along marks in recursive direct calls

judofyr committed:
Set default parser

judofyr committed:
DSL: Small fixes

judofyr committed:
Add another test

judofyr committed:
Update EBNF to use boundary helper

judofyr committed:
Introduce "boundary" helper instead of "plus_boundary"

judofyr committed:
Fix warnings

judofyr committed:
Update dependencies

judofyr committed:
Implement token look-ahead

judofyr committed:
Add CLI

judofyr committed:
P2: Support marks after tail call

judofyr committed:
P2: Tweak value propagation

judofyr committed:
P2: Remove `value` from return set

judofyr committed:
P2: final_marks -> final_values

judofyr committed:
Improve dot viz

judofyr committed:
Dot: Avoid circles for terminals

Since terminals can have quite long text

judofyr committed:
EBNF: Don't crash when the file ends with a prec rule

judofyr committed:
P2: Introduce action objects

judofyr committed:
Add partition refinement implementation

judofyr committed:
Tweak DOT output

- Detect format from filename
- Rename methods

judofyr committed:
bin/glush: Autoload Bundler locally

judofyr committed:
Remove duktape from Gemfile

It's not currently being used

judofyr committed:
Add PartRefine fixup

judofyr committed:
PartitionRefinement: Implement enumerable

judofyr committed:
ParseError: Set error message

judofyr committed:
P2: Fix before_marks in direct calls

judofyr committed:
Implement JavaScript generator

jsoroka commented:
> judofyr added 20 commits 20 days ago

+1

What a great surprise;  thank you for this work!

Michael35699 commented:
Hello, seeing as the last change is approx. 2 years ago, is this not being worked on anymore?

jsoroka commented:
I'm trying to understand and extend this project, but the lack of example EBNFs is frustrating my progress.  Can you add some sample EBNF files that produce a working javascript parser when run through the CLI, please?
